src/freertos_drivers/esp32/Esp32WS2812.cxx

```diff
@@ -0,0 +1,311 @@
+        {
+            // bytes are sent in MSB order
+            if (*psrc & (1 << i))
+            {
+                LOG(VERBOSE, "ESP-WS2812: %zu:%d 1", size, i);
+                pdest->val =  bit1.val;
+            }
+            else
+            {
+                LOG(VERBOSE, "ESP-WS2812: %zu:%d 0", size, i);
+                pdest->val =  bit0.val;
+            }
+            num++;
+            pdest++;
+        }
+        size++;
+        psrc++;
+    }
+
+    *translated_size = size;
+    *item_num = num;
+}
+
+// constructor
+Esp32WS2812::Esp32WS2812(const gpio_num_t output_pin,
+                         const rmt_channel_t channel,
+                         const size_t led_count)
+                       : OSThread(),
+                         pin_(output_pin),
+                         channel_(channel),
+                         ledCount_(led_count),
+                         ledBufferSize_(led_count * 3),
+                         sem_(0)
+{
+#if defined(CONFIG_IDF_TARGET_ESP32C3)
+    // ESP32-C3 has four RMT channels with the first two allocated for TX and
+    // second two for RX.
+    HASSERT(channel >= RMT_CHANNEL_0 && channel <= RMT_CHANNEL_1);
+#elif defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
+    // ESP32-S2 has four RMT channels.
+    // ESP32-S3 has eight RMT channels with the first four allocated for TX and
+    // remaining for RX.
+#elif defined(CONFIG_IDF_TARGET_ESP32)
+    // ESP32 has eight RMT channels.
+    HASSERT(channel >= RMT_CHANNEL_0 && channel <= RMT_CHANNEL_7);
+#endif
+    HASSERT(GPIO_IS_VALID_OUTPUT_GPIO(output_pin));
+}
+
+// destructor
+Esp32WS2812::~Esp32WS2812()
+{
+    free(ledDataBuffer_);
+}
+
+// initialize the perhipheral
+void Esp32WS2812::hw_init()
+{
+    LOG(VERBOSE, "ESP-WS2812(%d): Allocating %u bytes for %zu LEDs", channel_,
+        ledBufferSize_, ledCount_);
+    // Allocate the LED color data buffer
+    ledDataBuffer_ = (uint8_t *)calloc(1, ledBufferSize_);
+    HASSERT(ledDataBuffer_ != NULL);
+
+    rmt_config_t config = RMT_DEFAULT_CONFIG_TX(pin_, channel_);
+    // reconfigure the clock to 40Mhz
+    config.clk_div = 2;
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Configuring RMT using GPIO %d", channel_,
+        pin_);
+    ESP_ERROR_CHECK(rmt_config(&config));
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Installing RMT driver", channel_);
+    ESP_ERROR_CHECK(rmt_driver_install(channel_, 0, 0));
+
+    LOG(VERBOSE, "ESP-WS2812(%d): Installing RMT translator", channel_);
+    ESP_ERROR_CHECK(rmt_translator_init(channel_, ws2812_rmt_adapter));
+
+    if (timeZeroHighTicks == 0 || timeZeroLowTicks == 0 ||
+        timeOneHighTicks == 0 || timeOneLowTicks == 0)
+    {
+        // convert the number of microseconds for high/low signals to the
+        // number of RMT ticks required.
+        uint32_t counter_clk_hz = 0;
+        ESP_ERROR_CHECK(rmt_get_counter_clock(channel_, &counter_clk_hz));
+        float ratio = (float)counter_clk_hz / 1e9;
```

Comment:
you could get rid of the floating point here by taking clock_hz / 1000000 as ticks_per_usec. This is typically an integer (and if it isn't, we're only doing  1% clock error which the WS2812 will be OK with).

Reply:
Agreed, there is also a new led_strip driver component in IDF v4.4 that will simplify this code greatly. I'll test that and update for using that instead of carrying the code here provided it works correctly.

Reply:
Fixed